Lab4/src/ca/mcgill/ecse211/lab4/UltrasonicLocalizer.java

```diff
@@ -0,0 +1,253 @@
+ * {@value #filterControl} Class variable to be incremented in our filter to remove false negatives
+ * {@value #lastDistance} Class variable used in our filter to keep track of the last distance returned by the ultrasonic sensor
+ * {@value #locType} Class variable. Localization type that needs to be specified when an instance of the class is created
+ *
+ * 
+ * @author Maxime Bourassa
+ * @author Violet Wei
+ */
+public class UltrasonicLocalizer extends Thread {
+
+  private LocalizationType locType;
+  private static final int MAX_DISTANCE = 50;
+  private static final int SMALL_DISTANCE = 31;
+  private static final float WALL_DISTANCE = 30;
+  private static final float MARGIN_DISTANCE =(float) 0.5; 
+  private static int FILTER_OUT = 10;
+  private int filterControl;
+  private float lastDistance;
+
+  /**
+   * Constructor of the class. Takes in only the localization type 
+   * as everything else we need are constants from the Lab4 class
+   * @param locType
+   */
+  public UltrasonicLocalizer(LocalizationType locType) {
+    this.locType = locType;
+  }
+
+  /**
+   * Entry point of the thread. Will run either falling or
+   * rising edge localization. All the heavy lifting of the
+   * class is done in this method
+   */
+  public void run() {
+    //variable used to start an average of the values we get in our noise margin
+    boolean noiseZone = false;
+    //first angle before noise margin
+    double firstangle = 0;
+    //First angle at the end of the noise margin
+    double firstangle2 = 0;
+    //second angle before noise margin
+    double secondangle = 0;
+    //Second angle at the end of the noise margin
+    double secondangle2 = 0;
+    //If falling edge has been selected, continue
+    if(locType == LocalizationType.FALLING_EDGE) {
+      //Calls our navigation method to set
+      Nav.setSpeeds(SPEED);
+      //if the robot starts facing a wall, turn until there is no longer a wall
+      while (getFilter(MAX_DISTANCE)<=WALL_DISTANCE) {
+        //turns the robot in a clockwise direction
+        Nav.turnClockwise();
+      }
+      // When the robot sees a wall, record the angle at the beginning and at the end of the noise margin
+      while (true) {
+        Nav.turnClockwise();
+        //Used to detect the angle before the noise margin
+        if (!noiseZone && getFilter(MAX_DISTANCE) <= WALL_DISTANCE + MARGIN_DISTANCE) {
+          //get it from the odometer
+          firstangle = odometer.getTheta();
+          //we are now in the noise margin
+          noiseZone = true;
+          //records the angle at the end of the noise margin
+        } else if ( getFilter(MAX_DISTANCE) >= WALL_DISTANCE - MARGIN_DISTANCE && noiseZone) {
+          firstangle2 = odometer.getTheta();
+        } else if(noiseZone) {
+          //We passed through the noise zone, put the boolean variable back to false

```

Comment:
Useless comment, especially the second part of it. Anyone who knows basic coding can see that you're setting it to `false`.